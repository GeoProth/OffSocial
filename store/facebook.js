/* eslint-disable */
import axios from "axios";

const state = () => ({
  FB: {},
  UserInfo: {},
  UserEvents: [],
  albumsApi: {},
  albums: [],
  loggedIn: false,
  grams: []
});
const mutations = {
  SET_FB_API(state, api) {
    state.FB = api;
  },
  SET_USER_INFO(state, user) {
    state.UserInfo = user;
  },
  SET_USER_EVENTS(state, payload) {
    state.UserEvents = payload;
  },
  SET_ALBUMS(state, albums) {
    state.albums = albums;
  },
  SET_FB_LOGIN(state) {
    state.loggedIn = true;
  },
  SET_FB_LOGOUT(state) {
    state.loggedIn = false;
  },
  SET_GRAMS(state, grams) {
    state.grams = grams;
  }
};
const actions = {
  setSDK({ commit }, fbSDK) {
    commit("SET_FB_API", fbSDK);
  },
  async getUserInfo({ commit }) {
    /* eslint-disable no-console */
    await window.FB.api(
      "/me",
      "GET",
      { fields: "name, id, first_name, last_name, picture, email" },
      user => {
        console.log("User info: ", user);
        commit("SET_USER_INFO", user);
      }
    );
  },
  async getEvents({ commit }) {
    await window.FB.api(
      "/me/events",
      { fields: "name, place, description, start_time, end_time, rsvp_status" },
      response => {
        commit("SET_USER_EVENTS", response);
      }
    );
  },
  async getGrams({ commit }) {
    /* eslint-disable no-console */
    const {
      data: { data }
    } = await axios.get(
      "https://api.instagram.com/v1/users/self/media/recent/?",
      {
        params: {
          access_token: "AccessTokenHERE"
        }
      }
    );
    console.log("response instagram: ", data);

    const grams = data;
    commit("SET_GRAMS", grams);
  },
  async getAlbums({ commit }) {
    await window.FB.api(
      `me?fields=albums.fields(id,count,name,cover_photo.fields(id,name,source,picture),photos.fields(name,picture,source))`,
      response => {
        const albums = response.albums.data.map(album => ({
          count: album.count,
          cover_photo: album.cover_photo.source,
          id: album.id,
          name: album.name,
          photos: album.photos.data
        }));
        commit("SET_ALBUMS", albums);
      }
    );
    // commit('SET_ALBUMS', response.albums.data.map(album => { album.cover_photo = album.cover_photo.picture; }));
  },
  setUserLoggedIn({ commit }) {
    commit("SET_FB_LOGIN");
  },
  setUserLoggedOut({ commit }) {
    commit("SET_FB_LOGOUT");
  }
};
const getters = {
  FB: ({ FB }) => FB,
  UserInfo: ({ UserInfo }) => UserInfo,
  UserEvents: ({ UserEvents }) => UserEvents,
  Albums: ({ albums }) => albums,
  FBLoggedIn: ({ loggedIn }) => loggedIn,
  Grams: ({ grams }) => grams
};
export default {
  state,
  mutations,
  actions,
  getters
};
