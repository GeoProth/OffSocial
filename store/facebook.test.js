
import axios from "axios";
import facebook from "./facebook";

jest.mock("axios");
describe("Facebook store", () => {
  const mockFB = {
    api: jest.fn()
  };
  Object.defineProperty(window, "FB", {
    value: mockFB,
    writable: true
  });
  const ctx = {
    commit: jest.fn(),
    setState(state) {
      this.state = state;
    },
    resetState() {
      this.state = undefined;
    }
  };

  beforeEach(() => {
    axios.post.mockReset();
    axios.get.mockReset();
    axios.put.mockReset();
    axios.patch.mockReset();
    axios.delete.mockReset();

    facebook.actions.$axios = axios;
    facebook.getters.$axios = axios;
    facebook.mutations.$axios = axios;

    ctx.commit.mockReset();
    ctx.resetState();
    mockFB.api.mockReset();
  });
  test("has correct shape", () => {
    expect(facebook.state).toBeFunction();
    expect(facebook.mutations).toBeObject();
    expect(facebook.actions).toBeObject();
    expect(facebook.getters).toBeObject();
  });
  test("state has the correct shape", () => {
    const state = facebook.state();
    expect(state.FB).toBeObject();
    expect(state.UserInfo).toBeObject();
    expect(state.UserEvents).toBeArray();
    expect(state.albums).toBeArray();
    expect(state.grams).toBeArray();
  });
  // ACTIONS -----------------------------------------------------
  describe("Actions", () => {
    // Set FB SDK
    test("setSDK passes correct arguments", () => {
      facebook.actions.setSDK(ctx, { fizz: "buzz" });
      expect(ctx.commit).toHaveBeenCalledWith("SET_FB_API", { fizz: "buzz" });
    });
    // Set User Info
    test("setUserInfo passes correct arguments", async () => {
      mockFB.api.mockImplementation((...args) => args.map((arg) => arg instanceof Function ? arg({ fizz: "buzz" }) : null));
      await facebook.actions.getUserInfo(ctx, { fizz: "buzz" });
      expect(ctx.commit).toHaveBeenCalledWith("SET_USER_INFO", { fizz: "buzz" });
    });
    test("getEvents passes correct args", async () => {
      ctx.setState({});
      mockFB.api.mockImplementation((...args) => args.map((arg) => arg instanceof Function ? arg({ spam: "eggs" }) : null));
      await facebook.actions.getEvents(ctx);
      expect(mockFB.api).toBeCalled();
      expect(ctx.commit).toBeCalledWith("SET_USER_EVENTS", { spam: "eggs" });
    });

    test("getAlbums calls the facebook api", async () => {
      await facebook.actions.getAlbums(ctx);
      expect(mockFB.api).toBeCalled();
    });
    // set grams
    test("getGrams calls instagram api", async () => {
      await facebook.actions.getGrams(ctx);
      expect(mockFB.api).toBeCalled();
    });
  });

  // MUTATIONS ---------------------------------------------------
  describe("Mutations", () => {
    // Set FB ABI
    test("SET_FB_API adds the fb object to the state", () => {
      const mockState = {};
      facebook.mutations.SET_FB_API(mockState, { fizz: "buzz" });
      expect(mockState.FB).toContainAllKeys(["fizz"]);
    });
    // set User Info
    test("SET_USER_INFO adds the user object to the state", () => {
      const mockState = {};
      facebook.mutations.SET_USER_INFO(mockState, { fizz: "buzz" });
      expect(mockState.UserInfo).toContainAllKeys(["fizz"]);
    });
    // set Albums
    test("SET_ALBUMS adds the albums array to the state", () => {
      const mockState = {
        albums: []
      };
      const mockAlbum = {
        data: [{ fizz: "buzz" }]
      };
      facebook.mutations.SET_ALBUMS(mockState, mockAlbum);
      expect(mockState).toMatchSnapshot();
    });
    // Set User Events
    test("SET_USER_EVENTS adds the user Events array to the state", () => {
      const mockState = [];
      facebook.mutations.SET_USER_EVENTS(mockState, ["1", "2", "3"]);
      expect(mockState.UserEvents).toContainAllValues(["1", "2", "3"]);
    });
    // set Insta gram
    test("SET_GRAMS adds user instagram array to state", () => {
      const mockState = [];
      facebook.mutations.SET_GRAMS(mockState, ["1", "2", "3"]);
      expect(mockState.grams).toContainAllValues(["1", "2", "3"]);
    });
  });
  // GETTERS ------------------------------------------------------------
  describe("Getters", () => {
    // GET FB state
    test("FB Getter should return correct property", () => {
      const mockState = { FB: "spam" };
      expect(facebook.getters.FB(mockState)).toBe("spam");
    });
    // GET UserInfo state
    test("UserInfo Getter should return correct property", () => {
      const mockState = { UserInfo: "spam" };
      expect(facebook.getters.UserInfo(mockState)).toBe("spam");
    });
    // GET User Albums state
    test("UserAlbums getter should return correct array", () => {
      const mockState = { albums: ["1", "2", "3"] };
      expect(facebook.getters.Albums(mockState)).toEqual(["1", "2", "3"]);
    });
    // GET UserEvents state
    test("UserEvents getters should return correct array", () => {
      const mockState = { UserEvents: ["1", "2", "3"] };
      expect(facebook.getters.UserEvents(mockState)).toEqual(["1", "2", "3"]);
    });
    // GET Grams state
    test("grams getter shoudl return correct array", () => {
      const mockState = { grams: ["1", "2", "3"] };
      expect(facebook.getters.grams(mockState)).toEqual(["1", "2", "3"]);
    });
  });
});
