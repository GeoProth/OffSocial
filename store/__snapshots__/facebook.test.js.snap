// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Facebook store Mutations SET_ALBUMS adds the albums array to the state 1`] = `
Object {
  "albums": Object {
    "data": Array [
      Object {
        "fizz": "buzz",
      },
    ],
  },
}
`;
